apps/mark-scan/pat-device-input/src/pat_input_reader.rs

```diff
@@ -0,0 +1,231 @@
+use std::{fmt, io};
+use vx_logging::{log, EventId};
+
+use crate::pin::Pin;
+
+// Debian 12 will offset pin addresses by 512.
+// eg. the is_connected pin is 478 + 512 = 990
+// Even though prod is expected to need the offset, we use the constant
+// values below because
+// 1. These are the documented values
+// 2. Starting low and adding the offset instead of starting high and
+//    subtracting the offset means we don't have to worry about
+//    pin addresses < 0
+const POSSIBLE_PIN_OFFSET: u16 = 512;
+const IS_CONNECTED_PIN_ADDRESS: u16 = 478;
+const SIGNAL_A_PIN_ADDRESS: u16 = 481;
+const SIGNAL_B_PIN_ADDRESS: u16 = 476;
+
+pub struct PatInputReader<T: Pin> {
+    is_connected_pin: Option<T>,
+    signal_a_pin: Option<T>,
+    signal_b_pin: Option<T>,
+}
+
+impl<T: Pin> PatInputReader<T> {
+    pub fn new() -> Self {
+        Self {
+            is_connected_pin: Some(T::new(IS_CONNECTED_PIN_ADDRESS + POSSIBLE_PIN_OFFSET)),
```

Comment:
Agreed, I liked that the default (of using the offset) is explicit in `new` but it is weird to put that assumption in 2 separate places. Will update.

Reply:
Ah I remember now. In `test_connect_offset` we need `offset = 512` to fail so it tries `offset = 0`. We do this by making `reader.connect()` probe `self.is_connected_pin`. Then in the test we set `is_connected_pin` to a `MockPin` that fails on `probe()`. Dependency injection works here because `connect` assumes `new` has already set up `is_connected_pin` for its first attempt. If we make `is_connected_pin = None` to start and instead `connect()` just calls `T::new` for a function-scoped `Pin` to probe, then we need to be able to control the mock without having a reference to the `Pin`.

I'm not sure if there's a preferred way to do this without dependency injection. I could create a third implementer of `Pin`, eg `MockZeroOffsetPin`, whose `probe()` fails specifically for `address = 990` and have the test use a `PatInputReader<MockZeroOffsetPin>`. But that seems oddly specific. Or I could make a `PinFactory` layer that `reader` uses to make `Pins`, like (pseudocode):

```
trait PinFactory {
  create(&self) -> Pin;
}

impl RealPinFactory for PinFactory {
  fn create(&self, addr: u16) -> GpioPin {
    GpioPin::new(addr)
  }

struct TestPinFactory {
  addrs_that_fail: [u16]
}

impl TestPinFactory for PinFactory {
  fn create(&self, addr: u16) -> MockPin {
    if self.addrs_that_fail.contains(addr) {
      MockPin { address: addr, active: false, probe_error: Some(io::Error::new(...)) }
    }
    MockPin { address: addr, active: false, probe_error: None }
  }
}
```

but this fundamentally has the same clunkiness of creating one or more mock implementations for a single test. Will think more on how to do this tomorrow.